For #349 View Downloads

kglazko opened the pull request:
### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

Mugurell commented on app/src/main/java/org/mozilla/fenix/downloads/DownloadsFragment.kt:
```diff
@@ -0,0 +1,129 @@
+import org.mozilla.fenix.R
+import org.mozilla.fenix.ext.requireComponents
+
+class DownloadsFragment: Fragment() {
```
If we are to bring a new Fragment it would be best to have it use the MVI architecture from the start - https://github.com/mozilla-mobile/fenix/blob/master/docs/architecture-overview.md

`BookmarkFragment`, `HistoryFragment` or `ShareFragment` can be used as an inspiration since they also have to display a list of items but they are not fully respecting the architecture.
We can do better.

Mugurell commented on app/src/main/java/org/mozilla/fenix/downloads/DownloadsItemMenu.kt:
```diff
@@ -0,0 +1,30 @@
+package org.mozilla.fenix.downloads
+
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
nit: the license header should be at the very top of the file, above the package statement - https://developer.android.com/kotlin/style-guide#structure

Mugurell commented on app/src/main/java/org/mozilla/fenix/downloads/DownloadsItemMenu.kt:
```diff
@@ -0,0 +1,30 @@
+import org.mozilla.fenix.R
+import org.mozilla.fenix.library.LibraryItemMenu
+
+class DownloadsItemMenu(
```
Seems like this is not used?

Mugurell commented on app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt:
```diff
@@ -163,6 +172,7 @@ class HomeMenu(
                 BrowserMenuDivider(),
                 historyItem,
                 bookmarksItem,
+                downloadsItem,
```
I see UX recommended this new menu being above the history one - #349 (comment)

Mugurell commented on app/src/main/java/org/mozilla/fenix/library/history/viewholders/DownloadsListItemViewHolder.kt:
```diff
@@ -0,0 +1,106 @@
+package org.mozilla.fenix.library.history.viewholders
```
This seems like a leftover?

Mugurell commented on app/src/main/java/org/mozilla/fenix/tabtray/TabTrayViewHolder.kt:
```diff
@@ -72,7 +73,10 @@ class TabTrayViewHolder(
             thumbnailView.setImageBitmap(tab.thumbnail)
         } else {
-            imageLoader.loadIntoView(thumbnailView, ImageRequest(tab.id))
+            // Make sure we have the view's dimensions so we can load the image at the correct size
```
This seem like a merge of another commit?
Probably will not appear anymore after a rebase.

Mugurell commented on app/src/main/res/layout/component_downloads.xml:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?>
```
nit: missing license header

Mugurell commented on app/src/main/res/layout/component_downloads.xml:
```diff
@@ -0,0 +1,26 @@
+        android:layout_gravity="bottom|end"
+        android:layout_margin="16dp" />
+
+
```
nit: no need for two newlines.

Mugurell commented on app/src/main/res/layout/download_list_item.xml:
```diff
@@ -0,0 +1,75 @@
+<?xml version="1.0" encoding="utf-8"?>
```
nit: missing license header

Mugurell replied:
nit: needs reformatting.

Mugurell commented on app/src/main/res/layout/download_list_item.xml:
```diff
@@ -0,0 +1,75 @@
+        app:layout_constraintEnd_toEndOf="parent"
+        android:background="?neutralFaded"
+        android:visibility="gone"/>
+</LinearLayout>
```
nit: no newline at end of file

Mugurell commented on app/src/main/res/layout/download_list_item.xml:
```diff
@@ -0,0 +1,75 @@
+        android:textAppearance="?attr/textAppearanceListItem"/>
+
+    <View
+        android:id="@+id/separator"
```
If we need a divider we could have our own `DividerItemDecoration` set for the RecycleView, one that would easily know to not be set after the last item.

Mugurell commented on app/src/main/res/layout/download_list_item.xml:
```diff
@@ -0,0 +1,75 @@
+        android:layout_marginStart="16dp"
+        android:layout_marginEnd="16dp"
+        android:importantForAccessibility="no"
+        app:layout_constraintTop_toTopOf="parent"
```
The container is a LinearLayout so these attributes will have no effect.

Mugurell commented:
Sorry, I only now saw this is a draft.
But I hope I helped a bit with moving this forward.

kglazko commented:
![DownloadDemoGif](https://user-images.githubusercontent.com/1676235/90082191-0c858880-dcc4-11ea-97a7-fc52b9bbe879.gif)
A visual update

kglazko commented:
I updated this PR, adding Nightly feature flags

ekager commented on app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserToolbarController.kt:
```diff
@@ -385,6 +385,13 @@ class DefaultBrowserToolbarController(
                 )
             }
+
+            ToolbarMenu.Item.Downloads -> {
```
To avoid the flash on the browser, you can add 	`browserAnimator.captureEngineViewAndDrawStatically {` before navigating like we do for history, bookmarks, etc here

liuche commented:
In the future, if you're pushing something very contained like a feature flag, just put it as a separate commit. In general, separate commits (that tell a story) will make things easier to review. (And the commits will all be squashed at the end anyway)

kglazko commented:
@liuche Sorry, will do! Got confused and just rebased everything into one commit instead. @ekager Awesome added it! @Mugurell Thank you for the review, lots of great feedback I have implemented, for the onCreateView/onCreate I modeled directly after the existing precedent in HistoryFragment though.

kglazko commented on app/src/main/res/layout/download_list_item.xml:
```diff
@@ -0,0 +1,17 @@
+<!-- This Source Code Form is subject to the terms of the Mozilla Public
+   - License, v. 2.0. If a copy of the MPL was not distributed with this
+   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+<LinearLayout
```
I am modelling off of history_list_item exactly for this one, I'm not sure if we will have more children for the future but I wanted to keep it consistent to be safe.

kglazko commented on app/src/main/res/drawable/ic_download_default.xml:
```diff
@@ -0,0 +1,12 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
Added it but it doesn't show as updated on GH for some reason
